snforge_std/src/cheatcodes/events.cairo

```diff
@@ -29,50 +20,73 @@ struct Event {
     data: Array<felt252>
 }
 
-
-/// An event spy structure, along with the events collected so far in the test.
-/// `events` are mutable and can be updated with `fetch_events`.
+/// An event spy structure containing an offset marking from which event given Spy
+/// should fetch events.
 #[derive(Drop, Serde)]
 struct EventSpy {
-    _id: felt252,
-    events: Array<(ContractAddress, Event)>,
+    _event_offset: usize
+}
+
+/// A wrapper structure on an array of events to handle filtering smoothly.
+#[derive(Drop, Serde)]
+struct Events {
+    events: Array<(ContractAddress, Event)>
 }
 
-trait EventFetcher {
-    /// Allows to update the structs' events field, from the spied contracts
-    fn fetch_events(ref self: EventSpy);
+trait EventSpyTrait {
+    /// Fetches all events accounting for the offset given [`EventSpy`] has.
+    fn get_events(ref self: EventSpy) -> Events;
 }
 
-impl EventFetcherImpl of EventFetcher {
-    fn fetch_events(ref self: EventSpy) {
-        let mut output = handle_cheatcode(cheatcode::<'fetch_events'>(array![self._id].span()));
+impl EventSpyTraitImpl of EventSpyTrait {
+    fn get_events(ref self: EventSpy) -> Events {
+        let mut output = handle_cheatcode(
+            cheatcode::<'get_events'>(array![self._event_offset.into()].span())
```

Comment:
Feels like a waste to always fetch those and not cache them somehow 🤔

Reply:
It's an idea for the future i think though

Reply:
I meant something that would not be exposed to the user, for performance

Reply:
Added an issue #2217